In [1]:
import pandas as pd
import numpy as np

In [2]:
crawl = pd.read_csv('Androcrawl.csv')

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2882: DtypeWarning: Columns (27,28,29,30,31,32,33,35,36,37,38,39,40,41,42,43,44,46,47,48,49,50,51,52,54,55,98,99,101,103,109,110,111,112,117,118,119,123,124,125,127,128,130,131,132,133,134,135,136,137,139,140,141,142,143,144,145,147,149,150,151,154,155,188,189,195,198,199,200,202,204,207,210,211,212,213,215,216,217,218,220) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [7]:
list(crawl.columns)

["'Apk Name'",
 "'Dangerous Permission: ACCESS_SUPERUSER'",
 "'Dangerous Permission: BLUETOOTH_PRIVILEGED'",
 "'Dangerous Permission: BRICK'",
 "'Dangerous Permission: CHANGE_COMPONENT_ENABLED_STATE'",
 "'Dangerous Permission: CLEAR_APP_USER_DATA'",
 "'Dangerous Permission: DELETE_CACHE_FILES'",
 "'Dangerous Permission: DELETE_PACKAGES'",
 "'Dangerous Permission: DISABLE_KEYGUARD'",
 "'Dangerous Permission: FACTORY_TEST'",
 "'Dangerous Permission: INSTALL_PACKAGES'",
 "'Dangerous Permission: INJECT_EVENTS'",
 "'Dangerous Permission: INTERNAL_SYSTEM_WINDOW'",
 "'Dangerous Permission: KILL_BACKGROUND_PROCESSES'",
 "'Dangerous Permission: MASTER_CLEAR'",
 "'Dangerous Permission: MODIFY_PHONE_STATE'",
 "'Dangerous Permission: MOUNT_FORMAT_FILESYSTEM'",
 "'Dangerous Permission: MOUNT_UNMOUNT_FILESYSTEM'",
 "'Dangerous Permission: PROCESS_OUTGOING_CALLS'",
 "'Dangerous Permission: READ_LOGS'",
 "'Dangerous Permission: REBOOT'",
 "'Dangerous Permission: RECEIVE_BOOT_COMPLETED'",
 "'Dangerous 

In [4]:
col = list(crawl.columns)

### Lista de permissões:

In [5]:
permissions = []
for i in col:
    if "Permission" in i:
            permissions.append(i)

In [6]:
permissions

["'Dangerous Permission: ACCESS_SUPERUSER'",
 "'Dangerous Permission: BLUETOOTH_PRIVILEGED'",
 "'Dangerous Permission: BRICK'",
 "'Dangerous Permission: CHANGE_COMPONENT_ENABLED_STATE'",
 "'Dangerous Permission: CLEAR_APP_USER_DATA'",
 "'Dangerous Permission: DELETE_CACHE_FILES'",
 "'Dangerous Permission: DELETE_PACKAGES'",
 "'Dangerous Permission: DISABLE_KEYGUARD'",
 "'Dangerous Permission: FACTORY_TEST'",
 "'Dangerous Permission: INSTALL_PACKAGES'",
 "'Dangerous Permission: INJECT_EVENTS'",
 "'Dangerous Permission: INTERNAL_SYSTEM_WINDOW'",
 "'Dangerous Permission: KILL_BACKGROUND_PROCESSES'",
 "'Dangerous Permission: MASTER_CLEAR'",
 "'Dangerous Permission: MODIFY_PHONE_STATE'",
 "'Dangerous Permission: MOUNT_FORMAT_FILESYSTEM'",
 "'Dangerous Permission: MOUNT_UNMOUNT_FILESYSTEM'",
 "'Dangerous Permission: PROCESS_OUTGOING_CALLS'",
 "'Dangerous Permission: READ_LOGS'",
 "'Dangerous Permission: REBOOT'",
 "'Dangerous Permission: RECEIVE_BOOT_COMPLETED'",
 "'Dangerous Permission: STA

### Lista de Chamadas de API:

In [48]:
APICalls = []
for i in col:
  if "Api" in i:
    APICalls.append(i)

In [49]:
APICalls

["'Api Call: Landroid/content/Context;->startService'",
 "'Api Call: Landroid/telephony/TelephonyManager;->getSubscriberId'",
 "'Api Call: Landroid/telephony/TelephonyManager;->getDeviceId'",
 "'Api Call: Landroid/telephony/TelephonyManager;->getLine1Number'",
 "'Api Call: Landroid/telephony/TelephonyManager;->getSimSerialNumber'",
 "'Api Call: Landroid/telephony/TelephonyManager;->getSimOperatorName'",
 "'Api Call: Landroid/telephony/TelephonyManager;->getCellLocation'",
 "'Api Call: Landroid/telephony/cdma/CdmaCellLocation;->getSystemId'",
 "'Api Call: Landroid/telephony/SmsManager;->sendTextMessage'",
 "'Api Call: Landroid/content/Intent;->setDataAndType'",
 "'Api Call: Landroid/content/Intent;->setType'",
 "'Api Call: Landroid/app/ActivityManager;->getRunningServices'",
 "'Api Call: Landroid/app/ActivityManager;->getMemoryInfo'",
 "'Api Call: Landroid/app/ActivityManager;->restartPackage'",
 "'Api Call: Landroid/content/pm/PackageManager;->getInstalledPackages'",
 "'Api Call: Ljava

In [50]:
no_APICalls = []
for i in APICalls:
  if "Notification" in i:
    no_APICalls.append(i)

for i in no_APICalls:
  APICalls.remove(i) 

In [51]:
APICalls

["'Api Call: Landroid/content/Context;->startService'",
 "'Api Call: Landroid/telephony/TelephonyManager;->getSubscriberId'",
 "'Api Call: Landroid/telephony/TelephonyManager;->getDeviceId'",
 "'Api Call: Landroid/telephony/TelephonyManager;->getLine1Number'",
 "'Api Call: Landroid/telephony/TelephonyManager;->getSimSerialNumber'",
 "'Api Call: Landroid/telephony/TelephonyManager;->getSimOperatorName'",
 "'Api Call: Landroid/telephony/TelephonyManager;->getCellLocation'",
 "'Api Call: Landroid/telephony/cdma/CdmaCellLocation;->getSystemId'",
 "'Api Call: Landroid/telephony/SmsManager;->sendTextMessage'",
 "'Api Call: Landroid/content/Intent;->setDataAndType'",
 "'Api Call: Landroid/content/Intent;->setType'",
 "'Api Call: Landroid/app/ActivityManager;->getRunningServices'",
 "'Api Call: Landroid/app/ActivityManager;->getMemoryInfo'",
 "'Api Call: Landroid/app/ActivityManager;->restartPackage'",
 "'Api Call: Landroid/content/pm/PackageManager;->getInstalledPackages'",
 "'Api Call: Ljava

### Lista de Chamadas de Sistema:

In [18]:
systemCalls = []
for i in col:
  if "System Routine" in i:
    systemCalls.append(i)

In [19]:
systemCalls

["'Calls a System Routine: su'",
 "'Calls a System Routine: ls'",
 "'Calls a System Routine: loadjar'",
 "'Calls a System Routine: grep'",
 "'Calls a System Routine: /sh'",
 "'Calls a System Routine: /bin'",
 "'Calls a System Routine: pm install'",
 "'Calls a System Routine: /dev/net'",
 "'Calls a System Routine: insmod'",
 "'Calls a System Routine: rm'",
 "'Calls a System Routine: mount'",
 "'Calls a System Routine: root'",
 "'Calls a System Routine: /system'",
 "'Calls a System Routine: stdout'",
 "'Calls a System Routine: reboot'",
 "'Calls a System Routine: killall'",
 "'Calls a System Routine: chmod'",
 "'Calls a System Routine: stderr'",
 "'Calls a System Routine: ratc'"]

### Lista de Intenções:

In [44]:
intents = []
for i in col:
  if "Intent" in i:
    intents.append(i)

In [45]:
intents

["'Api Call: Landroid/content/Intent;->setDataAndType'",
 "'Api Call: Landroid/content/Intent;->setType'",
 "'Api Call: Landroid/content/Intent;->describeContents'",
 "'C2M Intent: c2dm.intent.RECEIVE'",
 "'C2M Intent: c2dm.intent.REGISTRATION'",
 "'C2M Intent: c2dm.intent.REGISTER'",
 "'Suspicious Intent Filter: AIRPLANE_MODE_CHANGED'",
 "'Suspicious Intent Filter: BOOT_COMPLETED'",
 "'Suspicious Intent Filter: CONFIGURATION_CHANGED'",
 "'Suspicious Intent Filter: BATTERY_CHANGED'",
 "'Suspicious Intent Filter: DEVICE_STORAGE_LOW'",
 "'Suspicious Intent Filter: DOCK_EVENT'",
 "'Suspicious Intent Filter: EXTERNAL_APPLICATIONS_AVAILABLE'",
 "'Suspicious Intent Filter: MANAGE_PACKAGE_STORAGE'",
 "'Suspicious Intent Filter: MEDIA_MOUNTED'",
 "'Suspicious Intent Filter: MY_PACKAGE_REPLACED'",
 "'Suspicious Intent Filter: NEW_OUTGOING_CALL'",
 "'Suspicious Intent Filter: PACKAGE_ADDED'",
 "'Suspicious Intent Filter: PACKAGE_CHANGED'",
 "'Suspicious Intent Filter: PACKAGE_DATA_CLEARED'",
 "'

In [46]:
no_intents = []
for i in intents:
  if "Api" in i:
    no_intents.append(i)

for i in no_intents:
  intents.remove(i) 

In [47]:
intents

["'C2M Intent: c2dm.intent.RECEIVE'",
 "'C2M Intent: c2dm.intent.REGISTRATION'",
 "'C2M Intent: c2dm.intent.REGISTER'",
 "'Suspicious Intent Filter: AIRPLANE_MODE_CHANGED'",
 "'Suspicious Intent Filter: BOOT_COMPLETED'",
 "'Suspicious Intent Filter: CONFIGURATION_CHANGED'",
 "'Suspicious Intent Filter: BATTERY_CHANGED'",
 "'Suspicious Intent Filter: DEVICE_STORAGE_LOW'",
 "'Suspicious Intent Filter: DOCK_EVENT'",
 "'Suspicious Intent Filter: EXTERNAL_APPLICATIONS_AVAILABLE'",
 "'Suspicious Intent Filter: MANAGE_PACKAGE_STORAGE'",
 "'Suspicious Intent Filter: MEDIA_MOUNTED'",
 "'Suspicious Intent Filter: MY_PACKAGE_REPLACED'",
 "'Suspicious Intent Filter: NEW_OUTGOING_CALL'",
 "'Suspicious Intent Filter: PACKAGE_ADDED'",
 "'Suspicious Intent Filter: PACKAGE_CHANGED'",
 "'Suspicious Intent Filter: PACKAGE_DATA_CLEARED'",
 "'Suspicious Intent Filter: PACKAGE_FIRST_LAUNCH'",
 "'Suspicious Intent Filter: PACKAGE_INSTALL'",
 "'Suspicious Intent Filter: PACKAGE_REMOVED'",
 "'Suspicious Intent

### Lista de todas as colunas de interesse: 

In [52]:
crawlResultList = permissions + APICalls + systemCalls +intents

In [53]:
crawlResultList

["'Dangerous Permission: ACCESS_SUPERUSER'",
 "'Dangerous Permission: BLUETOOTH_PRIVILEGED'",
 "'Dangerous Permission: BRICK'",
 "'Dangerous Permission: CHANGE_COMPONENT_ENABLED_STATE'",
 "'Dangerous Permission: CLEAR_APP_USER_DATA'",
 "'Dangerous Permission: DELETE_CACHE_FILES'",
 "'Dangerous Permission: DELETE_PACKAGES'",
 "'Dangerous Permission: DISABLE_KEYGUARD'",
 "'Dangerous Permission: FACTORY_TEST'",
 "'Dangerous Permission: INSTALL_PACKAGES'",
 "'Dangerous Permission: INJECT_EVENTS'",
 "'Dangerous Permission: INTERNAL_SYSTEM_WINDOW'",
 "'Dangerous Permission: KILL_BACKGROUND_PROCESSES'",
 "'Dangerous Permission: MASTER_CLEAR'",
 "'Dangerous Permission: MODIFY_PHONE_STATE'",
 "'Dangerous Permission: MOUNT_FORMAT_FILESYSTEM'",
 "'Dangerous Permission: MOUNT_UNMOUNT_FILESYSTEM'",
 "'Dangerous Permission: PROCESS_OUTGOING_CALLS'",
 "'Dangerous Permission: READ_LOGS'",
 "'Dangerous Permission: REBOOT'",
 "'Dangerous Permission: RECEIVE_BOOT_COMPLETED'",
 "'Dangerous Permission: STA

## Dataset Final:

In [73]:
crawl["'Detection Ratio'"].value_counts()

Benign       13279
Malicious     1446
Name: 'Detection Ratio', dtype: int64

In [78]:
crawlResult = crawl[crawlResultList]

In [79]:
crawlResult["'Detection Ratio'"] = crawl["'Detection Ratio'"]

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [80]:
crawlResult["'Detection Ratio'"] = crawlResult["'Detection Ratio'"].replace("Malicious",int(1))
crawlResult["'Detection Ratio'"] = crawlResult["'Detection Ratio'"].replace("Benign",int(0))

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [81]:
crawlResult

'Dangerous Permission: ACCESS_SUPERUSER'  \
0                                         False   
1                                         False   
2                                         False   
3                                         False   
4                                         False   
...                                         ...   
14721                                     False   
14722                                     False   
14723                                     False   
14724                                     False   
14725                                     False   

       'Dangerous Permission: BLUETOOTH_PRIVILEGED'  \
0                                             False   
1                                             False   
2                                             False   
3                                             False   
4                                             False   
...                                             ...   
14721                                         False   
14722                                         False   
14723                                         False   
14724                                         False   
14725                                         False   

       'Dangerous Permission: BRICK'  \
0                              False   
1                              False   
2                              False   
3                              False   
4                              False   
...                              ...   
14721                          False   
14722                          False   
14723                          False   
14724                          False   
14725                          False   

       'Dangerous Permission: CHANGE_COMPONENT_ENABLED_STATE'  \
0                                                  False        
1                                                  False        
2                                                  False        
3                                                  False        
4                                                  False        
...                                                  ...        
14721                                              False        
14722                                              False        
14723                                              False        
14724                                              False        
14725                                              False        

       'Dangerous Permission: CLEAR_APP_USER_DATA'  \
0                                            False   
1                                            False   
2                                            False   
3                                            False   
4                                            False   
...                                            ...   
14721                                        False   
14722                                        False   
14723                                        False   
14724                                        False   
14725                                        False   

       'Dangerous Permission: DELETE_CACHE_FILES'  \
0                                           False   
1                                           False   
2                                           False   
3                                           False   
4                                           False   
...                                           ...   
14721                                       False   
14722                                       False   
14723                                       False   
14724                                       False   
14725                                       False   

       'Dangerous Permission: DELETE_PACKAGES'  \
0                                        False   
1                                        False   
2                                  

In [82]:
crawlResult["'Detection Ratio'"].value_counts()


0.0    13279
1.0     1446
Name: 'Detection Ratio', dtype: int64

### Dataset resultante contém 134 features: permissões, chamadas de sistema, chamadas de API e intenções de 14726 amostras (13279 benignos + 1446 maliciosos)

In [83]:
crawlResult.to_csv("2CPI_Androcrawl.csv")

2CPI porque são 2 Calls, Permissions e Intents